### ISSUE TITLE
[post] mocking vs final - which to choose?

### ISSUE THREAD

**Issue by TomasVotruba** [TomasVotruba/tomasvotruba.cz#642]
People often chose wrong architecure just to make testing tool happy

- https://www.reddit.com/r/PHP/comments/ajdgx6/when_to_declare_classes_final/eevqmdo/
- https://github.com/dg/bypass-finals

**Comment by TomasVotruba** [comment]
It requires listener in PHPUnit actually! Super hard to find out :/

```xml
<phpunit bootstrap="vendor/autoload.php">
    <listeners>
        <listener class="Elasticr\PHPUnit\Listener\BypassFinalListener"/>
    </listeners>
</phpunit>
```

```php
<?php declare(strict_types=1);

namespace Elasticr\PHPUnit\Listener;

use DG\BypassFinals;
use PHPUnit\Framework\AssertionFailedError;
use PHPUnit\Framework\Test;
use PHPUnit\Framework\TestListener;
use PHPUnit\Framework\TestSuite;
use PHPUnit\Framework\Warning;

final class BypassFinalListener implements TestListener
{
    public function addError(Test $test, \Throwable $t, float $time): void
    {
    }

    public function addWarning(Test $test, Warning $e, float $time): void
    {
    }

    public function addFailure(Test $test, AssertionFailedError $e, float $time): void
    {
    }

    public function addIncompleteTest(Test $test, \Throwable $t, float $time): void
    {
    }

    public function addRiskyTest(Test $test, \Throwable $t, float $time): void
    {
    }

    public function addSkippedTest(Test $test, \Throwable $t, float $time): void
    {
    }

    public function startTestSuite(TestSuite $suite): void
    {
        BypassFinals::enable();
    }

    public function endTestSuite(TestSuite $suite): void
    {
    }

    public function startTest(Test $test): void
    {
    }

    public function endTest(Test $test, float $time): void
    {
    }
}
```

**Comment by TomasVotruba** [comment]
Resolved by #701
